integration/pubsub_test.go

```diff
@@ -227,6 +254,93 @@ func (suite *serviceBusSuite) duplicateDetectionTest(pub *publisher.Publisher, l
 	)
 }
 
+func (suite *serviceBusSuite) publishAndConcurrentReceiveMessage(testConfig publishReceiveTest, numOfEvents int) {
+	ctx := context.Background()
+	if testConfig.listenerOptions == nil {
+		testConfig.listenerOptions = []listener.Option{}
+	}
+	if testConfig.publisherOptions == nil {
+		testConfig.publisherOptions = []publisher.Option{}
+	}
+
+	events := []testEvent{}
+	for i := 0; i < numOfEvents; i++ {
+		j := i + 1
+		event := testEvent{
+			ID:    j,
+			Key:   fmt.Sprintf("key%d", j),
+			Value: fmt.Sprintf("value%d", j),
+		}
+		events = append(events, event)
+	}
+
+	testConfig.listenerOptions = append(testConfig.listenerOptions, listener.WithConcurrency(3))
+
+	// setup listener
+	gotMessage := make(chan string)
+	go func() {
+		err := testConfig.listener.Listen(
+			ctx,
+			checkConcurrentResultHandler(gotMessage),
+			testConfig.topicName,
+			testConfig.listenerOptions...,
+		)
+		if err != nil {
+			fmt.Printf("listener error: %s", err)
+		}
+	}()
+	// publish after the listener is setup
+	time.Sleep(5 * time.Second)
+	publishCount := numOfEvents
+	for i := 0; i < publishCount; i++ {
+		err := testConfig.publisher.Publish(
+			ctx,
+			events[i],
+			testConfig.publisherOptions...,
+		)
+		suite.NoError(err)
+	}
+	var msgstatus []string
+	expected := []string{
+		"recieved1",
+		"recieved2",
+		"recieved3",
+		"completed1",
+		"recieved4",
+		"completed2",
+		"recieved5",
+		"completed3",
+		"recieved6",
+		"completed4",
+		"completed5",
+		"completed6",
+	}
+
+	for {
+		select {
+		case status := <-gotMessage:
+			if status == "error" && testConfig.shouldSucceed {
+				suite.FailNow("unhappy response")
+			}
+			msgstatus = append(msgstatus, status)
+
+		case <-time.After(10 * time.Second):
+			if testConfig.shouldSucceed {
+				suite.FailNow("Test didn't finish on time %v", msgstatus)
+			}
+		}
+		if len(msgstatus) >= len(expected) {
+			break
+		}
+	}
+
+	assert.Equal(suite.T(), msgstatus, expected)
+
+	//time.Sleep(20 * time.Second)
```

Comment:
this is added to not close listener before it prints out the msgs, I think it's not need anymore as for the loop above blocks until we finish the checks.